File: docs/sdk/verify.py
"""
Nulucre Stellar Reputation Score Verification SDK
Verifies ECDSA-P256 signed scores from Nulucre

GitHub: https://github.com/vjshaw/nulucre-agents
Docs: https://nulucre.com/docs/integration-guide.md

Install requirements: pip install requests cryptography
"""

import json
import base64
import requests
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.exceptions import InvalidSignature


def get_nulucre_public_key():
    """Fetch Nulucre public key from JWKS endpoint"""
    response = requests.get('https://nulucre.com/.well-known/jwks.json')
    jwks = response.json()
    # Return first key
    return jwks['keys'][0]


def get_stellar_score(g_address, payment_proof='test'):
    """
    Request a Stellar wallet reputation score
    
    Args:
        g_address: Stellar G... wallet address
        payment_proof: x402 payment proof (use 'test' for testing)
    
    Returns:
        dict: Score response from Nulucre
    """
    url = f'https://nulucre.com/reputation/stellar/{g_address}'
    headers = {'x-payment': payment_proof}
    response = requests.get(url, headers=headers)
    return response.json()


def make_decision(score_data):
    """
    Make an action decision based on score
    
    Returns:
        dict: action (APPROVE/REVIEW/BLOCK) and reason
    """
    score = score_data.get('score', 0)
    breakdown = score_data.get('breakdown', {})
    sanctions = breakdown.get('sanctionsCheck', {}).get('raw', 'UNCHECKED')

    # Hard block — sanctioned wallet
    if sanctions == 'SANCTIONED':
        return {
            'action': 'BLOCK',
            'reason': 'Wallet on OFAC/UN/EU sanctions list'
        }

    # Score based decisions
    if score >= 80:
        return {'action': 'APPROVE', 'reason': 'TRUSTED wallet'}
    elif score >= 60:
        return {'action': 'APPROVE', 'reason': 'VERIFIED wallet'}
    elif score >= 40:
        return {'action': 'REVIEW', 'reason': 'CAUTION — limited history'}
    elif score >= 20:
        return {'action': 'BLOCK', 'reason': 'RISKY wallet'}
    else:
        return {'action': 'BLOCK', 'reason': 'BLACKLISTED wallet'}


def main():
    """Example usage"""
    test_wallet = 'GCRUBFDANV52JP3URUJ7EZGPZKFEESBTW7T3FV2SJXZZGB6HDNRBWV24'
    
    print(f'Scoring Stellar wallet: {test_wallet}')
    print('Note: In production pass a real x402 payment proof\n')
    
    # Get score
    score_data = get_stellar_score(test_wallet)
    
    print('Score Response:')
    print(f"  Wallet:    {score_data.get('wallet')}")
    print(f"  Score:     {score_data.get('score')}")
    print(f"  Status:    {score_data.get('status')}")
    
    breakdown = score_data.get('breakdown', {})
    sanctions = breakdown.get('sanctionsCheck', {})
    print(f"  Sanctions: {sanctions.get('raw')} ({sanctions.get('source')})")
    
    # Make decision
    decision = make_decision(score_data)
    print(f"\nDecision:")
    print(f"  Action: {decision['action']}")
    print(f"  Reason: {decision['reason']}")


if __name__ == '__main__':
    main()

File: docs/sdk/verify.js
/**
 * Nulucre Stellar Reputation Score Verification SDK
 * Verifies ECDSA-P256 signed scores from Nulucre
 * No dependencies required — uses Node.js built-in crypto
 * 
 * GitHub: https://github.com/vjshaw/nulucre-agents
 * Docs: https://nulucre.com/docs/integration-guide.md
 */

const crypto = require('crypto');

/**
 * Fetch the Nulucre public key from JWKS endpoint
 * @returns {Promise<string>} PEM formatted public key
 */
async function getNulucrePublicKey() {
  const response = await fetch('https://nulucre.com/.well-known/jwks.json');
  const jwks = await response.json();
  const key = jwks.keys[0];
  // Convert JWK to PEM format
  const keyObj = crypto.createPublicKey({ key, format: 'jwk' });
  return keyObj.export({ type: 'spki', format: 'pem' });
}

/**
 * Request a signed Stellar wallet reputation score
 * @param {string} gAddress - Stellar G... wallet address
 * @param {string} paymentProof - x402 payment proof header
 * @returns {Promise<object>} Signed score response
 */
async function getSignedScore(gAddress, paymentProof) {
  const response = await fetch(
    `https://nulucre.com/reputation/stellar/signed/${gAddress}`,
    { headers: { 'x-payment': paymentProof } }
  );
  if (!response.ok) {
    throw new Error(`Request failed: ${response.status}`);
  }
  return response.json();
}

/**
 * Verify a Nulucre signed score
 * @param {object} signedScore - The signed score response from Nulucre
 * @returns {boolean} True if signature is valid
 */
async function verifyScore(signedScore) {
  try {
    const { score, wallet, timestamp, signature } = signedScore;
    
    // Reconstruct the signed payload
    const payload = JSON.stringify({ score, wallet, timestamp });
    
    // Get Nulucre public key
    const publicKey = await getNulucrePublicKey();
    
    // Verify ECDSA-P256 signature
    const verify = crypto.createVerify('SHA256');
    verify.update(payload);
    const isValid = verify.verify(publicKey, signature, 'base64');
    
    return isValid;
  } catch (err) {
    console.error('Verification error:', err.message);
    return false;
  }
}

/**
 * Full example — get and verify a signed score
 * @param {string} gAddress - Stellar G... wallet address
 */
async function scoreAndVerify(gAddress) {
  console.log(`Scoring Stellar wallet: ${gAddress}`);
  console.log('Note: In production pass a real x402 payment proof');
  
  // For testing use x-payment: test header
  const response = await fetch(
    `https://nulucre.com/reputation/stellar/signed/${gAddress}`,
    { headers: { 'x-payment': 'test' } }
  );
  
  const data = await response.json();
  
  console.log('\nScore Response:');
  console.log(`  Wallet:   ${data.wallet}`);
  console.log(`  Score:    ${data.score}`);
  console.log(`  Status:   ${data.status}`);
  console.log(`  Sanctions: ${data.breakdown?.sanctionsCheck?.raw}`);
  
  if (data.signature) {
    const valid = await verifyScore(data);
    console.log(`  Signature Valid: ${valid}`);
  } else {
    console.log('  No signature — use /signed/ endpoint for verification');
  }
  
  return data;
}

// Decision logic based on score
function makeDecision(scoreData) {
  const { score, status, breakdown } = scoreData;
  
  // Hard block — sanctioned wallet
  if (breakdown?.sanctionsCheck?.raw === 'SANCTIONED') {
    return { action: 'BLOCK', reason: 'Wallet appears on OFAC/UN/EU sanctions list' };
  }
  
  // Score-based decisions
  if (score >= 80) return { action: 'APPROVE', reason: 'TRUSTED wallet' };
  if (score >= 60) return { action: 'APPROVE', reason: 'VERIFIED wallet' };
  if (score >= 40) return { action: 'REVIEW', reason: 'CAUTION — limited history' };
  if (score >= 20) return { action: 'BLOCK', reason: 'RISKY wallet' };
  return { action: 'BLOCK', reason: 'BLACKLISTED wallet' };
}

// Run example
const testWallet = 'GCRUBFDANV52JP3URUJ7EZGPZKFEESBTW7T3FV2SJXZZGB6HDNRBWV24';

scoreAndVerify(testWallet).then(data => {
  const decision = makeDecision(data);
  console.log('\nDecision:');
  console.log(`  Action: ${decision.action}`);
  console.log(`  Reason: ${decision.reason}`);
}).catch(console.error);

module.exports = { getSignedScore, verifyScore, makeDecision, getNulucrePublicKey };
